### PR TITLE
[Performance] BEVFormer: Build invariant encoder tensors at init

### LINKED ISSUE #55194: [BEVFormer] Precompute and reuse invariant encoder tensors
## Context

The encoder uses several tensors whose values depend only on model configuration, BEV grid shape, or other inputs shared by all encoder layers. Recreating and uploading those tensors in every layer or attention call repeats the same setup work.

## Problem

- [`reference_points_3d` is regenerated for every encoder forward](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_encoder.py#L420-L432), although it is determined by BEV shape, batch size, point-cloud range, and pillar configuration.
- [Each layer extracts and uploads the same 2D BEV reference points](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_encoder.py#L155-L158).
- [The same BEV shape tensor is recreated inside every layer iteration](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_encoder.py#L476-L483).
- [TSA creates and uploads `level_start_index=[0]` when it is not provided](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_temporal_self_attention.py#L157-L164).
- [MSDA uploads the same spatial-shape values during each call](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_ms_deformable_attention.py#L282-L284).

## Proposed direction

- Cache `reference_points_3d` by BEV shape, batch size, point-cloud range, and pillar configuration.
- Create and upload 2D BEV reference points once per encoder forward, then pass the same device tensor to every layer.
- Create BEV shape metadata once before the layer loop.
- Create constant TSA metadata during module initialization.
- Cache MSDA's device representation of spatial shapes by their values and reuse it across calls.
- Rebuild a cached tensor when any input that determines its value, shape, dtype, or device changes.

## Acceptance criteria

Criteria 3, 4 and 6 were reformulated after implementation: the original wording
assumed a keyed cache, while the implementation fixes geometry as constructor state
instead. Evidence for each in the status comment below.

- [x] Every cached tensor documents which inputs determine its value and when it must be rebuilt.
- [x] Profiling confirms that each tensor is constructed or uploaded only once per required lifetime, rather than once per layer or attention call.
- [x] Geometry is validated constructor state rather than a cache: BEV shape and spatial shapes are checked at construction and asserted against runtime tensors, batch size is the only runtime degree of freedom and is covered by tests, and dtype and device are fixed by construction.
- [x] Memory growth is bounded and reported: no keyed structure exists, so the persistent set does not grow with batch size, forward count, or call count, and it is reported by tensor count and identity.
- [x] Existing TSA, SCA, point-sampling, layer, and encoder tests pass at their configured PCC thresholds, including PCC >= 0.997 for the base encoder.
- [x] Profiles report device-operation count, host gap, and wall time. Persistent memory is covered by the criterion above; device kernel time is omitted because the run's numbers do not reconcile.

## References

- [Encoder PCC tests](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tests/pcc/test_encoder.py)
- [Prototype implementation](https://github.com/tenstorrent/tt-metal/commit/7d6f4433bb38e0782d0e9ec468354bb3bb384134)

### PR DESCRIPTION
**Pull request by ctr-mmicicTT**
### PR Category
Performance

### Summary
BEV geometry (`bev_h`/`bev_w`, `spatial_shapes`) becomes keyword-only constructor
state of the encoder, the layer, TSA, SCA and MSDA, validated once at construction.
Everything derived from it is built there too: the bs=1 3D reference-point grid,
its 2D BEV slice as a device tensor, `bev_shape`, and the MSDA sampling-offset
normalizer.

Removed from the per-layer / per-call path:
- the 2D BEV reference-point slice and its host→device upload, done in every layer;
- the `bev_shape` tensor, rebuilt inside the layer loop;
- the MSDA spatial-shapes upload and offset-normalizer construction, done on every
  attention call;
- TSA's `level_start_index=[0]` construction and upload — MSDA never reads it, so
  the argument is gone from the TSA and layer signatures;
- SCA's spatial-shapes consistency check, now asserted once per forward in the
  encoder instead of once per layer.

Per forward the encoder only widens the stored bs=1 grid to the runtime batch
(host `expand`, `ttnn.repeat` on device, deallocated after the layer loop) and
projects it to camera coordinates. Batch size stays a runtime argument, so one
instance serves a varying batch without reconstruction.

Closes #55194
Depends on #55640

### Performance

Single encoder layer, Blackhole P100, `nuscenes_base`, 100x100 BEV, `batch_size=1`.
Source: `models/experimental/bevformer/tests/perf/test_layer_perf.py::test_bevformer_layer_perf`
Region: Tracy `start`/`stop` signposts, post-warmup iterations only.

| Metric | Before (`5b1b911`) | After (`4284945`) | Delta |
| --- | --- | --- | --- |
| Dispatched ops | 128 | 113 | **-15 (-11.7%)** |
| Summed device kernels | 318.128 ms | 317.071 ms | -1.06 ms (-0.33%) |
| Op-to-op gaps | 0.872 ms | 0.817 ms | -0.055 ms (-6.3%) |
| Forward elapsed | 319.000 ms | 317.888 ms | -1.11 ms (-0.35%) |

Perf reports:
[tt_perf_report.csv before](https://github.com/user-attachments/files/32245714/tt_perf_report.csv)
[tt_perf_report.csv after](https://github.com/user-attachments/files/32245731/tt_perf_report.csv)

The measured win is dispatch count, not device time. The 15 removed dispatches
carry 554 us of device time in total, so elapsed moves within noise:

| OP Code | Before | After | Device time delta |
| --- | --- | --- | --- |
| `ConcatDeviceOperation` | 3 | 1 | -500.7 us |
| `TilizeWithValPaddingDeviceOperation` | 5 | 3 | -34.7 us |
| `ReshapeViewDeviceOperation` | 20 | 15 | -19.5 us |
| `UntilizeWithUnpaddingDeviceOperation` | 13 | 11 | -3.9 us |
| `SliceDeviceOperation` | 13 | 9 | +4.6 us |

The removals match the per-call work this PR moves to construction: the 2D BEV
reference-point slice and its upload, the `bev_shape` rebuild, the MSDA
spatial-shapes upload and offset normalizer, and TSA's `level_start_index`
upload. Each is cheap on device; the point is that none of them is dispatched
from inside the layer any more.

This is a single-layer measurement. The removed work was per-layer, so a
`num_layers=6` encoder is expected to drop proportionally more dispatches --
that scaling is an expectation, not measured here.

**Provenance caveat:** the commit labels above come from the report metadata and
do not identify a checkoutable state and no longer exists on any branch
(rebased away).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-mmicic/55194-bevformer-precompute-invariant-encoder-tensors)
### Unit Tests
https://github.com/tenstorrent/tt-metal/actions/runs/35232969373

### CODE REVIEW

**Comment by ctr-mmicicTT** [comment]
Unit tests: 
https://github.com/tenstorrent/tt-metal/actions/runs/34497592246
https://github.com/tenstorrent/tt-metal/actions/runs/34588543349

**Comment by ctr-mmicicTT** [comment]
Nightly Unit tests on Tier3 fails already because of the CI time budget: https://github.com/tenstorrent/tt-metal/actions/runs/34570204001/job/103172148083

That is one of the reasons it has been increased here.

File: models/experimental/bevformer/tt/tt_encoder.py [hunk 1]
```diff
@@ -119,15 +126,13 @@ def forward(
...
-        spatial_shapes=None,
-        bev_shape=None,
         level_start_index=None,
         prev_bev=None,
         shift=None,
-        reference_points_3d=None,
         reference_points_cam=None,
         bev_mask=None,
         rebatch_plan=None,
+        bev_reference_points=None,
```
[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_encoder.py:134]: Resolved in https://github.com/tenstorrent/tt-metal/commit/172969f3c95e23663b07ce65406eaf2a22c8f007

File: models/experimental/bevformer/tests/pcc/test_spatial_cross_attention.py [hunk 2]
```diff
@@ -33,25 +33,29 @@
 # Default Test Configuration
 PRINT_DETAILED_COMPARISON_FLAG = False
 
-# Maximum relative error allowed for any query row
-MAX_QUERY_ROW_REL_ERROR = 0.3
+# Empirical threshold for gatther-scatter row-routing regressions that global PCC can dilute; not an accuracy tolerance.
```
[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tests/pcc/test_spatial_cross_attention.py]: Resolved in https://github.com/tenstorrent/tt-metal/commit/172969f3c95e23663b07ce65406eaf2a22c8f007

File: models/experimental/bevformer/tt/tt_encoder.py [hunk 3]
```diff
@@ -347,6 +368,26 @@ def __init__(
         self.batch_first = batch_first
         self.z_cfg = z_cfg
         self.feedforward_channels = feedforward_channels
+        self.bev_h = bev_h
+        self.bev_w = bev_w
+        self.spatial_shapes = spatial_shapes
+        self.bev_shape = torch.tensor([[bev_h, bev_w]], dtype=torch.long)
+
+        # The grid itself is batch-independent: batch size only broadcasts the leading
+        # dimension. Keep the bs=1 grid and widen it per forward.
```
[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_encoder.py]: Resolved in https://github.com/tenstorrent/tt-metal/commit/172969f3c95e23663b07ce65406eaf2a22c8f007

**Review by njovanovicTT — APPROVED** [review]
